### PR TITLE
Added TOC links and Engineering section to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by ezaquarii**
Proposal for disucssion:

1. add clickable links for TOC
2. add engineering practices section providing some guidlines based on code reviews

This is just a prototype requiring more work to make sure that it reflects Nextcloud GmbH position.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -257,12 +262,95 @@ Similar to layout files, menu files should match the name of the component. For
...
+
+This combination allows us to benefit from automation when it provides most value, does not tie rest of the code
+to any specific framework and stimulates continuous code modernization through iterative refactoring of all minor
+elements.
+
+### Custom platform APIs
+
+TL;DR Avoid Android platform APIs.
+
+NextCloud Android application provides some replacements for native Android APIs to facilitate testing
```
Comment (AndyScherzinger) [inline CONTRIBUTING.md]:
```suggestion
Nextcloud Android application provides some replacements for native Android APIs to facilitate testing
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -257,12 +262,95 @@ Similar to layout files, menu files should match the name of the component. For
...
+* device hardware information
+* media playback
+* networking
+* logging
+
+we have something more suitable.
+
+Out transition to new APIs is a continuous process. Contributors might be asked by code reviewers to
+refrain from  using specific Android APIs considered problematic and to use Nextcloud APIs instead.
+In extreme cases we might decide to put specific feature on hold until we provide platform API
```
Comment (AndyScherzinger) [inline CONTRIBUTING.md]:
```suggestion
In extreme cases we might decide to put specific features on hold until we provide platform API
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -257,12 +262,95 @@ Similar to layout files, menu files should match the name of the component. For
...
+
+We recognize the importance of technical debt that can slow down development, make bufixing difficult and
+discourage future contributors.
+
+We are mindful of the [Broken Windows Theory](https://en.wikipedia.org/wiki/Broken_windows_theory) and we'd like
+actively promote and encourage contributors to apply The Scout's Rule: *"Always leave the campground cleaner than 
+you found it"*. Simple, little improvements will sum up and will be very appreciated by Nextcloud team.
+
+We also promise to actively support and mentor contributors that help us to improve code quality, as we understand
+that this process is challenging and require deep understanding of the application codebase.
```
Comment (AndyScherzinger) [inline CONTRIBUTING.md]:
```suggestion
that this process is challenging and requires deep understanding of the application codebase.
```

**Comment by AndyScherzinger** [comment]
Looks good to me, just some minor typo things. So should we flag it to review for @tobiasKaminsky to have a look?

**Comment by ezaquarii** [comment]
@tobiasKaminsky @AndyScherzinger 
Apart from DCO and few others that I'll address, I'd really like to have a through-out review and discussion, as this change touches project's culture. :-)

Initially I thought I'll just open an Isssue with discussion, but nah, I think it's easier to discuss a diff than essays.

Being mergable is just a side effect of this practice, but it shall still be regarded as a fertilizer for discussion rather than PR.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -257,12 +262,95 @@ Similar to layout files, menu files should match the name of the component. For
...
+* application preferences
+* background task scheduling
+* device hardware information
+* media playback
+* networking
+* logging
+
+we have something more suitable.
+
+Out transition to new APIs is a continuous process. Contributors might be asked by code reviewers to
```
Comment (tobiasKaminsky) [inline CONTRIBUTING.md]:
```suggestion
Our transition to new APIs is a continuous process. Contributors might be asked by code reviewers to
```

**Comment by tobiasKaminsky** [comment]
Fine with it.
Though the testing part will be hard for me ;-) I try to do it all the time, but still I find myself writing too much code without proper testing.

**Comment by tobiasKaminsky** [comment]
Wa have the codacy bot that fails if the threshold of tests is lowering.
Maybe we should enforce this…
(We can have of course always exceptions, but this way we force ourself in writing more tests)

**Comment by ezaquarii** [comment]
> Wa have the codacy bot that fails if the threshold of tests is lowering.
> Maybe we should enforce this…
> (We can have of course always exceptions, but this way we force ourself in writing more tests)

Yes, this would be a nice warning sign, although the code volume in legacy, untestable components still grows, the waiver will be frequently granted.

My main point was that we should encourage testability of **new** code. Old code modifications are different kind of animal - there is not much we can easily do there for now except continuous refactoring, so we should be realistic and pragmatic about it.

I'm not sure if the text captures that intention.

**Comment by AndyScherzinger** [comment]
@ezaquarii @tobiasKaminsky I'd be fine with merging the actual version so this gets back to master and is "published" we can still take it from there and further improve it (baby steps). What do you think?

**Comment by AndyScherzinger** [comment]
So again: @ezaquarii @tobiasKaminsky I'd be fine with merging the actual version so this gets back to master and is "published" we can still take it from there and further improve it (baby steps). What do you think?

**Pull request merged**